File: src/com/timepath/plaf/x/filechooser/BaseFileChooser.java
package com.timepath.plaf.x.filechooser;

import java.awt.Frame;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.logging.Logger;

/**
 *
 * @author timepath
 */
public abstract class BaseFileChooser {

    private static final Logger LOG = Logger.getLogger(BaseFileChooser.class.getName());

    protected Frame parent;

    public Frame getParent() {
        return parent;
    }

    public BaseFileChooser setParent(Frame parent) {
        this.parent = parent;
        return this;
    }

    protected String dialogTitle;

    public String getTitle() {
        if(dialogTitle == null) {
            return this.isSaveDialog() ? "Save" : "Open";
        }
        return dialogTitle;
    }

    public BaseFileChooser setTitle(String title) {
        this.dialogTitle = title;
        return this;
    }

    protected File directory;

    public File getDirectory() {
        return directory;
    }

    public BaseFileChooser setDirectory(File directory) {
        this.directory = directory;
        return this;
    }
    
    public BaseFileChooser setDirectory(String directoryPath) {
        if(directoryPath == null) {
            this.directory = null;
        } else {
            setDirectory(new File(directoryPath));
        }
        return this;
    }
    
    protected String file;

    public BaseFileChooser setFile(String file) {
        this.file = file;
        return this;
    }

    public String getFile() {
        return file;
    }

    public BaseFileChooser setFile(File file) {
        if(file != null) {
            setDirectory(file.getParentFile());
            setFile(file.getName());
        }
        return this;
    }
    
    protected String approveButtonText;

    public String getApproveButtonText() {
        return approveButtonText;
    }

    public BaseFileChooser setApproveButtonText(String approveButtonText) {
        this.approveButtonText = approveButtonText;
        return this;
    }
    
    public static enum DialogType {
        SAVE_DIALOG, OPEN_DIALOG
    }
    
    protected DialogType dialogType = DialogType.OPEN_DIALOG;

    public DialogType getDialogType() {
        return dialogType;
    }

    public BaseFileChooser setDialogType(DialogType dialogType) {
        this.dialogType = dialogType;
        return this;
    }
    
    public boolean isSaveDialog() {
        return dialogType == DialogType.SAVE_DIALOG;
    }
    
    public static enum FileMode {
        DIRECTORIES_ONLY, FILES_ONLY, FILES_AND_DIRECTORIES
    }
    
    protected FileMode fileMode = FileMode.FILES_ONLY;

    public FileMode getFileMode() {
        return fileMode;
    }

    public BaseFileChooser setFileMode(FileMode fileMode) {
        this.fileMode = fileMode;
        return this;
    }
    
    public boolean isDirectoryMode() {
        return this.getFileMode() == FileMode.DIRECTORIES_ONLY;
    }
    
    protected boolean multiSelectionEnabled;
    
    public boolean isMultiSelectionEnabled() {
        return multiSelectionEnabled;
    }
    
    public BaseFileChooser setMultiSelectionEnabled(boolean multiSelectionEnabled) {
        this.multiSelectionEnabled = multiSelectionEnabled;
        return this;
    }
    
    public BaseFileChooser() {
        
    }
    
    public BaseFileChooser(File currentDirectory) {
        setDirectory(currentDirectory);
    }
    
    public BaseFileChooser(String currentDirectoryPath) {
        setDirectory(currentDirectoryPath);
    }

    public abstract File[] choose() throws IOException;
    
    protected ArrayList<ExtensionFilter> filters = new ArrayList<ExtensionFilter>();
    
    public BaseFileChooser addFilter(ExtensionFilter ef) {
        filters.add(ef);
        return this;
    }
    
    public static class ExtensionFilter {
        
        public ExtensionFilter(String shortDescription, String... extensions) {
            this.description = shortDescription;
            this.extensions = Arrays.asList(extensions);
        }

        private final List<String> extensions;

        public List<String> getExtensions() {
            return extensions;
        }

        private final String description;

        public String getDescription() {
            return description;
        }
        
    }
    
}

File: src/main/kotlin/com/timepath/plaf/x/filechooser/BaseFileChooser.kt
package com.timepath.plaf.x.filechooser


import java.io.File
import java.io.IOException
import java.util.*
import java.util.logging.Logger
import java.awt.Frame

/**
 * @author TimePath
 */
public abstract class BaseFileChooser {
    public var parent: Frame? = null
        protected set
    protected var dialogTitle: String? = null
    public var directory: File? = null
        protected set
    protected var filters: MutableCollection<ExtensionFilter> = LinkedList()
    public var fileName: String? = null
        set
    var approveButtonText: String? = null
    public var dialogType: DialogType = DialogType.OPEN_DIALOG
        set
    var fileMode = FileMode.FILES_ONLY
    var multiSelectionEnabled: Boolean = false

    fun setDirectory(directoryPath: String?): BaseFileChooser {
        if (directoryPath == null) {
            directory = null
        } else {
            setDirectory(File(directoryPath))
        }
        return this
    }

    public fun setParent(parent: Frame?): BaseFileChooser {
        this.parent = parent
        return this
    }

    fun getTitle(): String {
        if (dialogTitle == null) {
            return if (isSaveDialog()) "Save" else "Open"
        }
        return dialogTitle!!
    }

    public fun setTitle(title: String?): BaseFileChooser {
        dialogTitle = title
        return this
    }

    protected fun isSaveDialog(): Boolean {
        return dialogType == DialogType.SAVE_DIALOG
    }

    public fun setDirectory(directory: File?): BaseFileChooser {
        this.directory = directory
        return this
    }

    public fun setFileName(fileName: String?): BaseFileChooser {
        this.fileName = fileName
        return this
    }

    public fun setFile(file: File?): BaseFileChooser {
        if (file != null) {
            setDirectory(file.getParentFile())
            setFileName(file.getName())
        }
        return this
    }

    fun setApproveButtonText(approveButtonText: String?): BaseFileChooser {
        this.approveButtonText = approveButtonText
        return this
    }

    public fun setDialogType(dialogType: DialogType): BaseFileChooser {
        this.dialogType = dialogType
        return this
    }

    protected fun isDirectoryMode(): Boolean {
        return fileMode == FileMode.DIRECTORIES_ONLY
    }

    public fun setFileMode(fileMode: FileMode): BaseFileChooser {
        this.fileMode = fileMode
        return this
    }

    protected fun isMultiSelectionEnabled(): Boolean {
        return multiSelectionEnabled
    }

    public fun setMultiSelectionEnabled(multiSelectionEnabled: Boolean): BaseFileChooser {
        this.multiSelectionEnabled = multiSelectionEnabled
        return this
    }

    throws(javaClass<IOException>())
    public abstract fun choose(): Array<File>?

    public fun addFilter(ef: ExtensionFilter): BaseFileChooser {
        filters.add(ef)
        return this
    }

    public enum class DialogType {
        SAVE_DIALOG
        OPEN_DIALOG
    }

    public enum class FileMode {
        DIRECTORIES_ONLY
        FILES_ONLY
        FILES_AND_DIRECTORIES
    }

    public class ExtensionFilter(public val description: String, vararg extensions: String) {

        private val extensions: List<String>

        {
            this.extensions = Arrays.asList<String>(*extensions)
        }

        public fun getExtensions(): List<String> {
            return Collections.unmodifiableList<String>(extensions)
        }
    }

    class object {

        private val LOG = Logger.getLogger(javaClass<BaseFileChooser>().getName())
    }
}
